homeassistant/components/aprilaire/sensor.py

```diff
@@ -0,0 +1,254 @@
+    ),
+    AprilaireSensorDescription(
+        key="air_cleaning_status",
+        translation_key="air_cleaning_status",
+        status_key=Attribute.AIR_CLEANING_AVAILABLE,
+        status_sensor_available_value=None,
+        status_sensor_exists_values=[1],
+        value_key=Attribute.AIR_CLEANING_STATUS,
+        value_fn=lambda _, value: AIR_CLEANING_STATUS_MAP.get(value),
+    ),
+    AprilaireSensorDescription(
+        key="fan_status",
+        translation_key="fan_status",
+        status_key=None,
+        status_sensor_available_value=None,
+        status_sensor_exists_values=[],
+        value_key=Attribute.FAN_STATUS,
+        value_fn=lambda _, value: FAN_STATUS_MAP.get(value),
+    ),
+)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up Aprilaire sensor devices."""
+
+    coordinator: AprilaireCoordinator = hass.data[DOMAIN][config_entry.unique_id]
+
+    assert config_entry.unique_id is not None
+
+    async_add_entities(
+        AprilaireSensor(coordinator, description, config_entry.unique_id)
+        for description in SENSOR_TYPES
+        if description.status_key is None
+        or description.status_sensor_exists_values is None
+        or coordinator.data.get(description.status_key)
+        in description.status_sensor_exists_values
+    )
+
+
+class AprilaireSensor(BaseAprilaireEntity, SensorEntity):
+    """Sensor entity for Aprilaire."""
+
+    entity_description: AprilaireSensorDescription
+
+    def __init__(
+        self,
+        coordinator: AprilaireCoordinator,
+        description: AprilaireSensorDescription,
+        unique_id: str,
+    ) -> None:
+        """Initialize a sensor for an Aprilaire device."""
+
+        self.entity_description = description
+
+        super().__init__(coordinator, unique_id)
+
+    @property
+    def available(self) -> bool:
+        """Return True if entity is available."""
+        if (
+            self.entity_description.status_key is None
+            or self.entity_description.status_sensor_available_value is None
+        ):
+            return True
+
+        if not super().available:
+            return False
+
+        return (
+            self.coordinator.data.get(self.entity_description.status_key)
+            == self.entity_description.status_sensor_available_value
+        )
+
+    @property
+    def native_unit_of_measurement(self) -> str | None:
+        """Return the unit of measurement."""
+        if (
+            self.entity_description.native_unit_of_measurement is None
+            and self.entity_description.state_class is not None
+        ):
+            if self._sensor_option_unit_of_measurement is UNDEFINED:
+                return self.hass.config.units.temperature_unit
+
+            return self._sensor_option_unit_of_measurement
+
+        return super().native_unit_of_measurement
```

Comment:
What does the Aprilaire device return? In what unit? Is this dynamic?

Looking at the util.py the device always returns celcius. Can you maybe give an example why the special rounding is needed?

Reply:
The device does always return celsius. The problem is when converting to Fahrenheit, the device itself uses away-from-zero midpoint rounding, whereas Python uses banker's rounding. This leads to inconsistent rounding in the case of values when the user is displaying values in Fahrenheit in HA.

Consider 22.5C which is 72.5F. With away-from-zero rounding, the device displays this as 73F. However with banker's rounding, HA would show this as 72F.

The solution as implemented is to have the sensor always report that it has the same native unit of measurement as it is configured to display in, and then handle the correct rounding when returning the native value.

Reply:
This actually reminds me that I need to do the same for the climate entity. This fix was made in the HACS repo after I started getting this merged into core, so I never made the change on the core side.

Reply:
Please let me know if you have any other questions about this